libp2p/protocols/pubsub/pubsub.nim

```diff
@@ -47,61 +48,74 @@ type
     handler*: seq[TopicHandler]
 
   PubSub* = ref object of LPProtocol
+    switch*: Switch                                 # the switch used to dial/connect to peers
     peerInfo*: PeerInfo                             # this peer's info
     topics*: Table[string, Topic]                   # local topics
-    peers*: Table[string, PubSubPeer]               # peerid to peer map
-    conns*: Table[PeerInfo, HashSet[Connection]]    # peers connections
+    peers*: Table[PeerID, PubSubPeer]               # peerid to peer map
     triggerSelf*: bool                              # trigger own local handler on publish
     verifySignature*: bool                          # enable signature verification
     sign*: bool                                     # enable message signing
     cleanupLock: AsyncLock
     validators*: Table[string, HashSet[ValidatorHandler]]
-    observers: ref seq[PubSubObserver] # ref as in smart_ptr
-    msgIdProvider*: MsgIdProvider      # Turn message into message id (not nil)
+    observers: ref seq[PubSubObserver]              # ref as in smart_ptr
+    msgIdProvider*: MsgIdProvider                   # Turn message into message id (not nil)
     msgSeqno*: uint64
+    lifetimeFut*: Future[void]                      # pubsub liftime future
 
-method handleDisconnect*(p: PubSub, peer: PubSubPeer) {.base.} =
+method unsubscribePeer*(p: PubSub, peerId: PeerID) {.base.} =
   ## handle peer disconnects
   ##
-  
-  if not(isNil(peer)) and peer.peerInfo notin p.conns:
-    trace "deleting peer", peer = peer.id
-    peer.onConnect.fire() # Make sure all pending sends are unblocked
-    p.peers.del(peer.id)
-    trace "peer disconnected", peer = peer.id
 
-    # metrics
-    libp2p_pubsub_peers.set(p.peers.len.int64)
+  trace "unsubscribing pubsub peer", peer = $peerId
+  if peerId in p.peers:
+    p.peers.del(peerId)
 
-proc onConnClose(p: PubSub, conn: Connection) {.async.} =
-  try:
-    let peer = conn.peerInfo
-    await conn.closeEvent.wait()
+  libp2p_pubsub_peers.set(p.peers.len.int64)
 
-    if peer in p.conns:
-      p.conns[peer].excl(conn)
-      if p.conns[peer].len <= 0:
-        p.conns.del(peer)
+proc broadcast*(p: PubSub,
+                sendPeers: HashSet[PubSubPeer] | seq[PubSubPeer] | array[1, PubSubPeer],
```

Comment:
We don't need the generic here as @arnetheduck noted, a `seq` should do just fine. This should also be private, I don't see a reason to expose it, maybe tests?

Reply:
Well, `array[1, PubSubPeer]` allows 0 heap allocations in the case of a single peer, also `HashSet[PubSubPeer] | seq[PubSubPeer]` is needed cos we consume both with hash and seq...

Seriously what is wrong with using generics?

Reply:
the use of generics has costs - in the case of async, these costs are significant in the sense that async itself introduces a lot of bloat that is multiplied for each generic type - in this particular case, there's hardly to gain by the generics: the sendPeers data will be copied over and over again as it travels through the async layers and closures - there are tens of heap allocations here already that async itself introduces, in addition to the return value etc. put in this perspective, it's questionable whether the bloat price is worth the difference in number of heap allocations.

Reply:
but whatever, it really doesn't matter that much either way - what matters is more the semantic juggling of a single send vs an aggegation of multiple sends